sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetConverter.scala

```diff
@@ -426,6 +423,33 @@ private[parquet] class CatalystPrimitiveConverter(
     parent.updateLong(fieldIndex, value)
 }
 
+/**
+ * A `parquet.io.api.PrimitiveConverter` that converts Parquet Binary to Catalyst String.
+ * Supports dictionaries to reduce Binary to String conversion overhead.
+ *
+ * Follows pattern in Parquet of using dictionaries, where supported, for String conversion.
+ *
+ * @param parent The parent group converter.
+ * @param fieldIndex The index inside the record.
+ */
+private[parquet] class CatalystPrimitiveStringConverter(parent: CatalystConverter, fieldIndex: Int)
+  extends CatalystPrimitiveConverter(parent, fieldIndex) {
+
+  private var dict: Array[String] = null
+
+  override def hasDictionarySupport: Boolean = true
+
+  override def setDictionary(dictionary: Dictionary):Unit =
+    dict = (for (i <- 0 to dictionary.getMaxId)
```

Comment:
random esoteric scala detail: `foreach` on a `Range` is actually just about as good as a while loop.  this **only** applies to `Range`, though, nothing else.  more [here](https://github.com/scala/scala/commit/4cfc633fc6).  I stumbled upon this a while back, if you are interested in reading more: http://imranrashid.com/posts/scala-loop-profiling/